aep/general/0155/aep.md

```diff
@@ -0,0 +1,154 @@
+  // This request is only idempotent if `idempotency_key` is provided.
+  //
+  // This key will be honored for at least one hour after the first time it is
+  // seen by the server.
+  //
+  // The key is restricted to 36 ASCII characters. A random UUID is recommended.
+  aep.api.IdempotencyKey idempotency_key = 3 [
+      (aep.api.field_info).minimum_lifetime = { seconds: 3600 }
+  ];
+}
+```
+
+- [`aep.api.IdempotencyKey`][] has a `key` and a `first_sent` timestamp.
+
+  - `key` is simply a unique identifier.
+
+- Providing an idempotency key **must** guarantee idempotency.
+
+  - If a duplicate request is detected, the server **must** return one of:
+
+    - A response equivalent to the response for the previously successful
+      request, because the client most likely did not receive the previous
+      response.
+    - An error indicating that the `first_sent` field of the idempotency key is
+      invalid or cannot be honored (expired, in the future, or differs from a
+      previous `first_sent` value with the same `key`).
+    - An error, if returning an equivalent response is not possible.
+
+      For example, if a resource was created, then deleted, and then a
+      duplicate request to create the resource is received, the server **may**
+      return an error if returning the previously created resource is not
+      possible.
+
+  - APIs **should** honor idempotency keys for at least an hour.
+    - When using protocol buffers, idempotency keys that are UUIDs **must** be
+      annotated with a minimum lifetime using the extension
+      [`(aep.api.field_info).minimum_lifetime`][].
+
+- The `idempotency_key` field **must** be provided on the request message to
+  which it applies (and it **must not** be a field on resources themselves).
+
+  - The `first_sent` field can be used by API servers to determine if a key is
+    expired. API servers **must** reject requests with expired keys, and
+    **must** reject requests with keys that are in the future. When feasible,
+    API servers **should** reject requests that use the same `key` but have a
+    different `first_sent` timestamp.
+  - The `key` field **must** be able to be a UUID, and **may** allow UUIDs to
+    be the only valid format. The format restrictions for idempotency keys
+    **must** be documented.
+
+- Idempotency keys **should** be optional.
+
+### Stale success responses
+
+In some unusual situations, it may not be possible to return an identical
+success response. For example, a duplicate request to create a resource may
+arrive after the resource has not only been created, but subsequently updated;
+because the service has no other need to retain the historical data, it is no
+longer feasible to return an identical success response.
+
+In this situation, the method **may** return the current state of the resource
+instead. In other words, it is permissible to substitute the historical success
+response with a similar response that reflects more current data.
```

Comment:
Copying from https://github.com/aep-dev/aep.dev/pull/77#discussion_r1410793374 :

This behavior seems problematic, because such a response does not actually correlate with the request! What if the resource has subsequently been deleted, or made inaccessible? And even if the common case, **how is a client to detect that they have received such a response?**

Rather than silently replacing the no longer available appropriate response, I think it would be best to communicate that information. In HTTP terms, it would presumably be a [303 See Other](https://www.rfc-editor.org/rfc/rfc9110.html#section-15.4.4) response with a Location field identifying the URL to use for getting current state. And having provided that signal, it would then be possible _as an optimization_ to also include that representation along with a [Content-Location](https://www.rfc-editor.org/rfc/rfc9110.html#section-8.7) field, e.g.
```
HTTP/1.1 303 See Other
…
Location: /path/to/resource
Content-Location: /path/to/resource

<current representation>
```